torch/csrc/jit/passes/batch_mm.cpp

```diff
@@ -260,8 +260,134 @@ void BatchMMBlock(Block* block) {
+        other_side_inputs.reserve(num_other_side_inputs);
+        for (auto it = stack.end() - num_other_side_inputs; it != stack.end(); ++it) {
+          other_side_inputs.push_back(std::move(*it).toTensor());
+        }
+        drop(stack, num_other_side_inputs);
+        pop(stack, side_input);
+
+        auto any_other_input = other_side_inputs[0];
+        if (have_same_shape(other_side_inputs) && shape_is_fast_for_side(other_side_inputs[0])) {
+          auto other_side_input = at::cat(other_side_inputs, single_side == Side::LHS ? 1 : 0);
+          auto mm_out = single_side == Side::LHS ? side_input.mm(other_side_input) : other_side_input.mm(side_input);
+          auto outputs = at::chunk(mm_out, num_other_side_inputs, /*dim=*/single_side == Side::LHS ? 1 : 0);
+          stack.insert(stack.end(), std::make_move_iterator(outputs.begin()),
+                                    std::make_move_iterator(outputs.end()));
+        } else {
+          if (single_side == Side::LHS) {
+            for (at::Tensor& other : other_side_inputs) {
+              stack.emplace_back(side_input.mm(other));
+            }
+          } else {
+            for (at::Tensor& other : other_side_inputs) {
+              stack.emplace_back(other.mm(side_input));
+            }
+          }
+        }
+
+        return 0;
+      };
+    })
+});
+
+std::pair<std::vector<Node*>, std::vector<Node*>> gatherIndependentMMUses(Value *value) {
+  static const auto postprocess = [](std::vector<Node*> mms) {
+    if (mms.size() == 0) {
+      return mms;
+    }
+    std::sort(mms.begin(), mms.end(), [](Node* n, Node* m) { return n->isBefore(m); });
+    // Filter out dependent MMs. This algorithm might do very badly if e.g. you have
+    // a lot of independent MMs, that depend on the first one, but I doubt this will
+    // be a common scenario.
+    for (size_t i = 0; i < mms.size(); ++i) {
+      if (mms[i] == nullptr) continue;
+      for (size_t j = i + 1; j < mms.size(); ++j) {
+        if (mms[j] == nullptr) continue;
+        if (!mms[j]->couldMoveBeforeTopologically(mms[i])) {
+          mms[j] = nullptr;
+        }
+      }
+    }
+    return filter(mms, [](Node *n) { return n != nullptr; });
+  };
+
+  Block * block = value->node()->owningBlock();
+  std::vector<Node*> lhses; // Will contain nodes where value is used as an lhs
+  std::vector<Node*> rhses; // Like above, but rhs
+  for (Use u : value->uses()) {
+    if (u.user->owningBlock() == block &&
+        u.user->matches("aten::mm(Tensor self, Tensor mat2) -> Tensor")) {
+      if (u.offset == 0 && u.user->inputs()[1] != value) {
+        lhses.push_back(u.user);
+      } else if (u.offset == 1 && u.user->inputs()[0] != value) {
+        rhses.push_back(u.user);
+      }
+    }
+  }
+  return std::make_pair(postprocess(lhses), postprocess(rhses));
+}
+
+void BatchMMSide(Block * block) {
+  // NB: 8 is the current loop unrolling factor
+  static constexpr size_t how_many_is_many = 8;
```

Comment:
Would it be better to expose the loop unrolling factor from the loop unrolling pass instead of recording 8 here and there?

Reply:
That's a good point